toolchain/check/import_ref.cpp

```diff
@@ -224,6 +333,145 @@ class ImportRefResolver {
+    auto& new_scope = context_.name_scopes().Get(new_class.scope_id);
+    const auto& old_scope = import_ir_.name_scopes().Get(import_class.scope_id);
+    // Push a block so that we can add scoped instructions to it, primarily for
+    // textual IR formatting.
+    context_.inst_block_stack().Push();
+    for (auto [entry_name_id, entry_inst_id] : old_scope.names) {
+      CARBON_CHECK(
+          new_scope.names
+              .insert({GetLocalNameId(entry_name_id),
+                       context_.AddPlaceholderInst(SemIR::ImportRefUnused{
+                           import_ir_id_, entry_inst_id})})
+              .second);
+    }
+    new_class.body_block_id = context_.inst_block_stack().Pop();
+
+    if (import_class.base_id.is_valid()) {
+      new_class.base_id = base_const_id.inst_id();
+      // Add the base scope to extended scopes.
+      auto base_inst_id = context_.types().GetInstId(
+          context_.insts()
+              .GetAs<SemIR::BaseDecl>(new_class.base_id)
+              .base_type_id);
+      const auto& base_class = context_.classes().Get(
+          context_.insts().GetAs<SemIR::ClassType>(base_inst_id).class_id);
+      new_scope.extended_scopes.push_back(base_class.scope_id);
+    }
+    CARBON_CHECK(new_scope.extended_scopes.size() ==
+                 old_scope.extended_scopes.size());
+  }
+
+  auto TryResolveTypedInst(SemIR::ClassDecl inst, SemIR::InstId inst_id,
+                           bool incomplete_type) -> SemIR::ConstantId {
+    const auto& import_class = import_ir_.classes().Get(inst.class_id);
+
+    SemIR::ConstantId class_const_id = SemIR::ConstantId::Invalid;
+    // On the first pass, there's no incomplete type; start by adding one for
+    // any recursive references.
+    if (!incomplete_type) {
+      class_const_id = MakeIncompleteClass(inst_id, import_class);
+      // If there's only a forward declaration, we're done.
+      if (!import_class.object_repr_id.is_valid()) {
+        return class_const_id;
+      }
+      // This may not be needed because all constants might be ready, but we do
+      // it here so that we don't need to track which work item corresponds to
+      // this instruction.
+      work_stack_.back().incomplete_type = true;
+    }
+
+    CARBON_CHECK(import_class.object_repr_id.is_valid())
+        << "Only reachable when there's a definition.";
+
+    // Load constants for the definition.
+    auto initial_work = work_stack_.size();
+
+    auto object_repr_const_id = GetLocalConstantId(import_class.object_repr_id);
+    auto base_const_id = import_class.base_id.is_valid()
+                             ? GetLocalConstantId(import_class.base_id)
+                             : SemIR::ConstantId::Invalid;
+
+    if (HasNewWork(initial_work)) {
+      return SemIR::ConstantId::Invalid;
+    }
+
+    // On the first pass, we build the incomplete type's constant above. On the
+    // second pass, we need to fetch it.
```

Comment:
Maybe a bit more reminder of the connection between the specific bool variable and the passes?

```suggestion
    // On the first pass, we build the incomplete type's constant above. If we
    // get here on a subsequent pass we need to fetch the one we built in the
    // first pass.
```

Reply:
Done